### DISCUSSION TITLE
How do I replace nodes on the syntax tree?

### DISCUSSION THREAD

**Discussion by NMSAzulX** [dotnet/roslyn#54321] [Q&A] [answered]
In my scene, I found something in the syntax tree that needs to be replaced, such as "show();" needs to be replaced with "arg.show();" But I tried all kinds of methods and failed. Who can tell me what to do?
```C#
var model = com.GetSemanticModel(tree);
foreach(var node in nodes)
{
   SymbolInfo s = model.GetSymbolInfo(node);
   if (s.Symbol==null)
   {
          //how to create new node???
          //node = show
          //node = arg.show
          root = root.ReplaceNode(node, newNode);
   }
           
}
```

**Comment by alrz** [comment, marked as answer]
You should create a SimpleMemberAccessExpression for this.
```cs
// IdentifierNameSyntax name = 'show'
// ExpressionSytnax receiver = 'arg'
var newNode = SyntaxFactory.MemberAccessExpression(SyntaxKind.SimpleMemberAccessExpression, receiver, name);
var newRoot = root.Replace(name, newNode)
```
Note this probably won't work in a loop because once you've rewritten the root, the original nodes won't be a part of the new tree. I'd suggest using SyntaxEditor for this kind of rewriting.

**Reply by NMSAzulX** [reply]
Thank you very much. Yes it happened in a loop!

**Reply by NMSAzulX** [reply]
@alrz  I got errors from "semantiModel.GetDiagnostics()" .   But how to find the error node in SyntaxTree? 
```C#
foreach (var error in semantiModel.GetDiagnostics())
 {
     //I've been looking for it for a long time. The error doesn't contains syntaxNode . 
    
 }
```

**Reply by CyrusNajmabadi** [reply]
The error contains a location which contains a span.  This tells you where the problem is in the syntax tree.  This may or may not relate to a particular node (for example, it coudl cover two nodes, or a portion fo a single node if the issuer desired that).

**Reply by NMSAzulX** [reply]
@CyrusNajmabadi Thanks!
